llvm/lib/Target/RISCV/RISCVGatherScatterLowering.cpp

```diff
@@ -310,18 +307,31 @@ bool RISCVGatherScatterLowering::matchStridedRecurrence(Value *Index, Loop *L,
   }
   case Instruction::Mul: {
     Start = Builder.CreateMul(Start, SplatOp, "start");
-    Step = Builder.CreateMul(Step, SplatOp, "step");
     Stride = Builder.CreateMul(Stride, SplatOp, "stride");
     break;
   }
   case Instruction::Shl: {
     Start = Builder.CreateShl(Start, SplatOp, "start");
-    Step = Builder.CreateShl(Step, SplatOp, "step");
     Stride = Builder.CreateShl(Stride, SplatOp, "stride");
     break;
   }
   }
 
+  // Adjust the step value after its definition if it's an instruction.
+  if (auto *StepI = dyn_cast<Instruction>(Step))
+    Builder.SetInsertPoint(*StepI->getInsertionPointAfterDef());
```

Comment:
If the Step is a loop invariant in an earlier basic block, does this put the new step farther away then necessary. Should we only do this if the Step instruction is in the loop?

Reply:
Is it possible it puts it somewhere that SplatOp doesn't dominate?

Reply:
Good catch in both respects.  We need to insert this at the end of the preheader if Step is outside the loop.

Reply:
Good eye, should be fixed now.